### ISSUE TITLE
Add support for (and a small example of) indexing proto xrefs. [phab:D1093]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1442]
_Created by **zarko** at 2016-10-06 18:00:50_: 

This diff adds an example proto indexer (and Bazel support for
the same). It also includes a metadata handler for protoc's
metadata format. Finally, it provides an example script that shows
how to tie protobuf definitions in a .proto to uses in a .cc.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-06 20:12:19_:

Any divergence from the upstream/internal proto_library rules makes import substantially more difficult.  Any proto_library targets which actually use this will have to be omitted from the import (which is fine, just a heads up).

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-06 20:12:19_:

Can we avoid providing protected data members for subclasses to use directly?  At a minimum, this should be a thoroughly documented member function.  I'm not a big fan of having interface + partial implementation in one class, but it might be the best compromise here.

Also, it seems like minor changes could make this a constructor parameter of the specific Support implementations, rather than a possibly-unused member.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-06 20:12:19_:

GeneratedCodeInfo::Annotation?

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-06 20:12:19_:

`const auto&` ?

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-06 20:12:19_:

Does FIS take ownership of the fd?

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-06 20:12:19_:

This doesn't seem like something which should be hard-coded.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-06 20:12:19_:

There's no good reason for the `mapped_type` to be a pair, rather than a struct with properly named fields.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-06 20:12:19_:

Is there any particular reason this and Breadcrumb are nested types?  They seem to have enough behavior to warrant testing.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-06 20:12:19_:

probably want -e as well

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-06 20:12:19_:

Why have this script build all of these things rather than making it a sh_binary with runtime dependencies?

**Comment by kythe** [comment]
_**pludemann** commented inline (context lost during migration) at 2016-10-06 23:56:44_:

set -o errexit is easier to understand than -e

**Comment by kythe** [comment]
_**zarko** wrote at 2016-10-10 20:47:35_:

  - Address comments.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-11 16:39:26_:

std::function seemed fine for this purpose, rather than a one-function interface, my objection was just with mixing behavior and interface in an abstract base class.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-11 16:39:26_:

This is still hard-coded, just in a constant.  Can it not be a flag or parameter?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-10-11 23:39:18_:

It can't be a constructor parameter because metadata support is configured before the VFS is initialized and the VFS is required to look up VNames.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-10-11 23:39:18_:

Not unless you call `SetCloseOnDelete(true)`.

Note that if `file_stream.GetErrno() != 0`, the documentation specifies `the stream is broken and all subsequent operations will fail`, and I'm interpreting `Close` as an operation.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-10-11 23:39:18_:

`Breadcrumb` exists only to present an RAII-style interface to managing `path_`. I've moved some of the members out as they do seem to use `thiz_` too much.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-10-11 23:39:18_:

It wouldn't work with `bazel run` because it needs to invoke Bazel to build the `kindex`.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-10-11 23:39:18_:

This might end up doing less allocation in the end.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-10-11 23:39:18_:

Oh hey these comments didn't get published. Sorry about that.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-10-11 23:41:40_:

  - Address comments.
  - Address comments.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-10-13 18:21:37_:

ping

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-13 19:46:55_:

Sure, but looking at the code, that doesn't appear to necessarily be the case, just how it's organized now.  It's certainly possible that this formulation is the cleanest, I'm just unconvinced.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-13 19:46:55_:

This doesn't appear to be a case where that matters, nor is it necessarily true (particularly when using shared_ptr<>).  And it's definitely less readable (imo) than just using std::function.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-13 19:46:55_:

GeneratedCodeInfo::Annotation

**Comment by kythe** [comment]
_**shahms** wrote at 2016-10-13 19:46:55_:

A few more commands wrt VNameLookup

**Comment by kythe** [comment]
_**zarko** wrote at 2016-10-13 23:20:33_:

  - Address comments.
  - Address comments.
  - Address comments.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-14 22:28:36_:

So it occurs to me that I think my main complaint is that this class definition is too long and the inline nested classes as well as all of the inline member functions all contribute to that and make it hard to get a good idea of the class interface, members, etc.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-14 22:28:36_:

`std::vector<int> nodes(loc.path().begin(), loc.path().end());`  ?

Or:

```
paths_.emplace(std::vector<int>(loc.path().begin(), loc.path().end()),
               &loc);
```

If you want to get fancy, of course (don't think it's worth it):
```
paths_.emplace(
  std::piecewise_construct_t,
  std::forward_as_tuple(loc.path().begin(), loc.path.end()),
  &loc);
```

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-14 22:28:36_:

Not particularly a big fan of storing things that are only valid for the lifetime of a couple of function calls in a member variable, why is this not passed into `IndexDescriptor()`?  It actually looks like this member is only used in this function, why store it in a member at all?

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-14 22:28:36_:

This is a bit confusing.

 start = (0, 1)
  end = (0, 2) or (2, 3)

Can you add a comment as to why that is?

**Comment by kythe** [comment]
_**shahms** wrote at 2016-10-14 22:28:36_:

Just a couple more comments, looking at the code more closely.  I think my earlier "de-nest" comments were actually around overall length (and concomitant complexity) of the outer class.  It might be sufficient to just define some of the function out-of-line rather than doingitallinline.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-10-20 20:06:06_:

ping

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-20 21:51:32_:

`*vname = record->vname;`

(`CopyFrom()` is polymorphic and hazardous).

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-20 21:51:32_:

Move this into a separate function? `bool ReadFile(path, std::string*)`

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-10-20 21:51:32_:

false

**Comment by kythe** [comment]
_**zarko** wrote at 2016-10-20 22:05:22_:

  - Address comments.
  - Address comments.
  - Address comments.
  - Address comments.
  - Address comments.

**Comment by kythe** [comment]
_**shahms** wrote at 2016-10-20 22:08:24_:

Thanks for patience!

**Comment by kythe** [comment]
_**shahms** accepted this revision at 2016-10-20 22:08:24_

**Comment by kythe** [comment]
_**zarko** closed via commit 47021a5c2d048db1be3a22a227305a9332b0b3fe at 2016-10-20 22:09:15_
